lib/spack/spack/spec.py

```diff
@@ -1007,6 +1007,14 @@ def __init__(self, spec_like=None,
         # have package.py files for.
         self._normal = normal
         self._concrete = concrete
+
+        # Normalize external modules, to ensure they are never set from
+        # anything that is not a built-in Python list. This is needed
+        # because when hashes are computed there can be discrepancies
+        # depending on the type use to represent a sequence and round trip
+        # to YAML representation. See #18338
+        if external_modules:
+            external_modules = list(external_modules)
```

Comment:
Can we do this conversion in `package_prefs.spec_externals`? We could do type enforcement here but the conversion makes more sense where we pass the values in.

Reply:
The reasoning for putting the change here is that it's the Spec object that requires this protection to avoid issues like in #18338. If we solve it where we pass the values in we might be more exposed to regression in the future and might need to chase other places where we pass values in.

Reply:
While I think that the general principle of accommodating and formatting values as they are read in can improve robustness, this is crossing the line and getting too specific: reading through the `__init__` function and most of Spec, the concerns of YAML are isolated to a few functions - it's a bit jarring to read about the pitfalls of YAML representations in the middle of `Spec.__init__`. While Python is not strongly typed, we nevertheless have an implied interface for what is being passed in and in my opinion the real issue is that we violated that interface by passing a non-conformant value.

Like I said, if you want to do type checking here, I think that would be appropriate: we should probably enforce that the incoming type is a basic Python `list` (e.g. rather than a YAML sequence). I think the actual conversion to that list should occur elsewhere.

It's true that users could simply assign `Spec.external_modules`, but we can generally control that by adding `setter` methods.

Reply:
Apologies @scheibelp I missed the part in your first comment where you were suggesting doing also type enforcement in `__init__`. I'll do the change tomorrow and update docstrings where necessary.

Reply:
I gave some more thoughts to this and to your comment above and ended up moving the type conversion in `Spec.to_node_dict` as that method seems to be called by all hash-related functions. In this way:

1. We don't risk raising on Spec construction when checking the type of `external_modules` in a strict way
1. `Spec.__init__` is left untouched, so there won't be details there about this very specific YAML issue

What worried me when I was modifying `Spec.__init__` to have a line like:
```python
if external_modules:
    assert isinstance(external_modules, list), err_msg
```
is that the constraint above is really too tight and we'll eventually see errors only at runtime if something is not compliant. To state it otherwise, I was concerned that a modification like that might cause more troubles than it solve - and it would be there just for YAML conversion and hashes like you said. Since the code around hashing is well factored and all methods ultimately call `Spec.to_node_dict` it seemed better to move the conversion  and comment there.

Reply:
I prefer that `external_modules` be a "usable" type starting with `__init__` (I'd rather not keep `external_modules` in some "flawed" format until the Spec is written out).

I'm wondering if you would find it acceptable to try a modified version of your original approach: do the type conversion in `__init__` but handle it with a separate function - something like:

```
class Spec(...)

  @staticmethod
  format_module_list(module_list):
    """
    Given a module list, possibly read from a configuration file, return an
    object that serializes to a consistent YAML string before/after round-trip
    serialization to/from a Spec dictionary (stored in JSON format): when read in,
    the module list may contain YAML formatting that is discarded (non-essential)
    when stored as a Spec dictionary; we take care in this function to discard such
    formatting such that the Spec hash does not change before/after storage in JSON.
    """
```

Reply:
Done in 6fea35c